chain/store/splitstore.go

```diff
@@ -0,0 +1,334 @@
+	"github.com/filecoin-project/lotus/chain/types"
+	bstore2 "github.com/filecoin-project/lotus/lib/blockstore"
+)
+
+const CompactionThreshold = 5 * build.Finality
+
+var baseEpochKey = dstore.NewKey("baseEpoch")
+
+type SplitStore struct {
+	baseEpoch abi.ChainEpoch
+	curTs     *types.TipSet
+
+	cs *ChainStore
+	ds dstore.Datastore
+
+	hot  bstore2.Blockstore
+	cold bstore2.Blockstore
+
+	snoop TrackingStore
+	sweep TrackingStore
+
+	compacting bool
+}
+
+type TrackingStore interface {
+	Put(cid.Cid, abi.ChainEpoch) error
+	PutBatch([]cid.Cid, abi.ChainEpoch) error
+	Get(cid.Cid) (abi.ChainEpoch, error)
+	Delete(cid.Cid) error
+	Has(cid.Cid) (bool, error)
+	Keys() (<-chan cid.Cid, error)
+}
+
+var _ bstore2.Blockstore = (*SplitStore)(nil)
+
+// Blockstore interface
+func (s *SplitStore) DeleteBlock(cid cid.Cid) error {
+	// afaict we don't seem to be using this method, so it's not implemented
+	return errors.New("DeleteBlock not implemented on SplitStore; don't do this Luke!")
+}
+
+func (s *SplitStore) Has(cid cid.Cid) (bool, error) {
+	has, err := s.hot.Has(cid)
+
+	if err != nil {
+		return false, err
+	}
+
+	if has {
+		return true, nil
+	}
+
+	return s.cold.Has(cid)
+}
+
+func (s *SplitStore) Get(cid cid.Cid) (blocks.Block, error) {
+	blk, err := s.hot.Get(cid)
+
+	switch err {
+	case nil:
+		return blk, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.Get(cid)
+
+	default:
+		return nil, err
+	}
+}
+
+func (s *SplitStore) GetSize(cid cid.Cid) (int, error) {
+	size, err := s.hot.GetSize(cid)
+
+	switch err {
+	case nil:
+		return size, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.GetSize(cid)
+
+	default:
+		return 0, err
+	}
+}
+
+func (s *SplitStore) Put(blk blocks.Block) error {
+	epoch := s.curTs.Height()
+	err := s.snoop.Put(blk.Cid(), epoch)
+	if err != nil {
+		return err
+	}
+
+	return s.hot.Put(blk)
+}
+
+func (s *SplitStore) PutMany(blks []blocks.Block) error {
+	err := s.hot.PutMany(blks)
+	if err != nil {
+		return err
+	}
+
+	epoch := s.curTs.Height()
+
+	batch := make([]cid.Cid, 0, len(blks))
+	for _, blk := range blks {
+		batch = append(batch, blk.Cid())
+	}
+
+	return s.snoop.PutBatch(batch, epoch)
+}
+
+func (s *SplitStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	chHot, err := s.hot.AllKeysChan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	chCold, err := s.cold.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer cancel()
+
+		for _, in := range []<-chan cid.Cid{chHot, chCold} {
+			for cid := range in {
+				select {
+				case ch <- cid:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+func (s *SplitStore) HashOnRead(enabled bool) {
+	s.hot.HashOnRead(enabled)
+	s.cold.HashOnRead(enabled)
+}
+
+func (s *SplitStore) View(cid cid.Cid, cb func([]byte) error) error {
+	err := s.hot.View(cid, cb)
+	switch err {
+	case bstore.ErrNotFound:
```

Comment:
`ErrNotFound` is in `bstore2` (aka `lib/blockstore`) so you can use just it.
In essence, `lib/blockstore` is our adjusted version of `go-ipfs-blockstore`.

Reply:
aaah, nice; will do.

Reply:
done.